gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/TileSourceEntity.java

```diff
@@ -41,6 +40,7 @@
 
   @CopyAnnotations
   @NonNull
+  @PrimaryKey
```

Comment:
It's generally safer to use artificial ids (`id`) as PK, and to declare other unique columns as unique if necessary. If there's no specific reason for making this PK, can we move it back to `id`?

Reply:
Sorry, I was under the impression we'd need to use this as the PK if we want `insertOrUpdate` to work correctly and replace tiles with incremented area counts--will setting URL to Unique achieve the same?

Reply:
for reference 

> The implementation of the method will update its parameters in the database if they already exists (checked by primary keys). If they don't already exists, this option will not change the database.

https://developer.android.com/reference/androidx/room/Update

Reply:
@gino-m (just to ensure it winds up in your notifications! ...I can never tell w/ github)

Reply:
I think what you want here is the following in the DAO:
@Query("UPDATE tile_source SET reference_count=:newCount WHERE id = :id")
void updateReferenceCount(String id, int newCount);

You probably could also do the increment directly in SQL, but that might be overkill here.
Note that we've been following the convention that the table or entity is named after what each row describes, in this case "tile_source" instead of "tile_sources'.

Reply:
We can do that, but we'd have to change our tile source retrieval logic. At the moment it roughly looks like this:

1. Fetch the basemap JSON from the URL specified in the project
2. Use the JSON parser to parse tile sources that intersect with the selected area
3. Construct a tile source for each source
4. Download tiles sources and add them to the DB.

Step 3. is the problematic one. We generate a new UID every time we determine intersecting tiles, so, if a user adds an area that overlaps with an existing area, we can't rely on UIDs to update the reference counts of existing tiles, since these are uniquely determined at JSON parsing time. Thus assuming tile source A exists in the db and has and equivalent URL and extent as tile B (newly parsed), the UIDs for each of them will still be unique thus any updates to the DB that rely on a UID primary key will fail and instead fallback to inserting a new entry (for the same. tile). 

Hopefully that explanation makes sense, we can still keep UIDs as the primary key if that's better, we'd just need to add an additional DB call at JSON parsing time to eliminate disregard parsed tiles whose URLs already collide w/ those in the db (since this indicates the reference count should be incremented). This also puts us into a a weird situation in which it more than likely makes the most sense to perform count incrementation in the JSON parser...which seems to be a bad crossing of boundaries imo.

Hopefully this explanation makes sense. Happy to walk through the code as well if its easier.

Reply:
Hi @scolsen sorry I didn't see this before - I think we resolved this in our VCs, please cmiiw!